### ISSUE TITLE
ICC generates lots of unnecessary casts

### ISSUE THREAD

**Issue by Muon** [p12tic/libsimdpp#49]
When compiling with ICC 16.0, some bad overloads are selected for assignment (not sure if compiler bug?). This test code:

```c++
#define SIMDPP_ARCH_X86_SSE2
#include <simdpp/simd.h>
#include <cstdio>

int main()
{
	using namespace simdpp;

	uint32x4 v1 = make_ones<uint32x4>();
	v1 = v1 << 3;

	std::printf("%u\n", reduce_add(v1));

	return 0;
}
```

compiles to 

```asm
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\i_shift_l.h 
00FE1451  pcmpeqd     xmm1,xmm1  
00FE1455  pslld       xmm1,3  
--- C:\Users\Mak\Documents\bitpacker\test\simdpp_test1.cpp ---------------------
00FE145A  or          dword ptr [esp+80h],8000h  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE1465  movdqa      xmmword ptr [esp+90h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\test\simdpp_test1.cpp ---------------------
00FE146E  ldmxcsr     dword ptr [esp+80h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE1476  movaps      xmm2,xmmword ptr [esp+90h]  
00FE147E  movaps      xmmword ptr [esp+0A0h],xmm2  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE1486  movdqa      xmm0,xmmword ptr [esp+0A0h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\i_reduce_add.h 
00FE148F  movdqa      xmmword ptr [esp+90h],xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int8x16.h ------
00FE1498  movdqa      xmmword ptr [esp+0B0h],xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE14A1  movaps      xmm1,xmmword ptr [esp+0B0h]  
00FE14A9  movaps      xmmword ptr [esp+80h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\move_l.h -
00FE14B1  movdqa      xmm0,xmmword ptr [esp+80h]  
00FE14BA  psrldq      xmm0,8  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE14BF  movdqa      xmmword ptr [esp+0A0h],xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE14C8  movaps      xmm1,xmmword ptr [esp+0A0h]  
00FE14D0  movaps      xmmword ptr [esp+0B0h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\empty_expr.h ---
00FE14D8  movaps      xmm0,xmmword ptr [esp+0B0h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\i_reduce_add.h 
00FE14E0  movdqa      xmm1,xmmword ptr [esp+90h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\expr\i_add.h --
00FE14E9  paddd       xmm1,xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE14ED  movdqa      xmmword ptr [esp+80h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE14F6  movaps      xmm2,xmmword ptr [esp+80h]  
00FE14FE  movaps      xmmword ptr [esp+0A0h],xmm2  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE1506  movdqa      xmm0,xmmword ptr [esp+0A0h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\i_reduce_add.h 
00FE150F  movdqa      xmmword ptr [esp+90h],xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int8x16.h ------
00FE1518  movdqa      xmmword ptr [esp+0B0h],xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE1521  movaps      xmm1,xmmword ptr [esp+0B0h]  
00FE1529  movaps      xmmword ptr [esp+80h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\move_l.h -
00FE1531  movdqa      xmm0,xmmword ptr [esp+80h]  
00FE153A  psrldq      xmm0,4  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE153F  movdqa      xmmword ptr [esp+0B0h],xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE1548  movaps      xmm1,xmmword ptr [esp+0B0h]  
00FE1550  movaps      xmmword ptr [esp+0A0h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\empty_expr.h ---
00FE1558  movaps      xmm0,xmmword ptr [esp+0A0h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\i_reduce_add.h 
00FE1560  movdqa      xmm1,xmmword ptr [esp+90h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\expr\i_add.h --
00FE1569  paddd       xmm1,xmm0  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE156D  movdqa      xmmword ptr [esp+80h],xmm1  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\cast.inl ------
00FE1576  movaps      xmm2,xmmword ptr [esp+80h]  
00FE157E  movaps      xmmword ptr [esp+0B0h],xmm2  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\types\int32x4.h ------
00FE1586  movdqa      xmm0,xmmword ptr [esp+0B0h]  
--- C:\Users\Mak\Documents\bitpacker\lib\libsimdpp\simdpp\detail\insn\i_reduce_add.h 
00FE158F  movdqa      xmmword ptr [esp+90h],xmm0
```

**Comment by p12tic** [comment]
Have you compiled with full optimization on? The non-optimized code almost always has this kind of unnecessary data movement form stack to registers and back.

**Comment by Muon** [comment]
Yeah, checked with both `/Ox` and `/O3`.

**Comment by p12tic** [comment]
No longer happens as of 40cada321886fa976. Likely fixed by f5a9d754d5b6de.
